File: src/server/trading/indicators/PearsonCorrelation/PearsonCorrelation.ts
import { TIndicatorShortCandle } from '../types';

export class PearsonCorrelation {
  private calculate(x: number[], y: number[]) {
    const n = x.length;

    if (n < 2) {
      console.warn(
        'PearsonCorrelation: prices series have less than 2 observations:',
        x.length,
        y.length,
      );

      return null;
    }

    // Рассчитываем суммы
    let sumX = 0;
    let sumY = 0;
    let sumXY = 0;
    let sumX2 = 0;
    let sumY2 = 0;

    for (let i = 0; i < n; i++) {
      sumX += x[i];
      sumY += y[i];
      sumXY += x[i] * y[i];
      sumX2 += x[i] * x[i];
      sumY2 += y[i] * y[i];
    }

    // Рассчитываем корреляцию Пирсона
    const numerator = n * sumXY - sumX * sumY;
    const denominator = Math.sqrt((n * sumX2 - sumX * sumX) * (n * sumY2 - sumY * sumY));

    if (isNaN(numerator) || isNaN(denominator)) {
      return 0;
    }

    // Избегаем деления на ноль
    if (denominator === 0) {
      return 0;
    }

    return numerator / denominator;
  }

  public correlationByPrices(candlesA: TIndicatorShortCandle[], candlesB: TIndicatorShortCandle[]) {
    const minLength = Math.min(candlesA.length, candlesB.length);
    if (minLength < 2) {
      console.warn(
        'PearsonCorrelation: prices series have less than 2 observations:',
        candlesA.length,
        candlesB.length,
      );

      return null;
    }

    const pricesA = candlesA.slice(-minLength).map((candle) => candle.close);
    const pricesB = candlesB.slice(-minLength).map((candle) => candle.close);

    return this.calculate(pricesA, pricesB);
  }

  public rollingCorrelationByPrices(
    candlesA: TIndicatorShortCandle[],
    candlesB: TIndicatorShortCandle[],
    window: number = 100,
  ) {
    const minLength = Math.min(candlesA.length, candlesB.length);

    if (minLength < window) {
      return [];
    }

    const result: { timestamp: number; value: number | null }[] = [];

    for (let i = window; i < minLength; i++) {
      const seriesA = candlesA.slice(i - window, i);
      const seriesB = candlesB.slice(i - window, i);

      result.push({
        timestamp: Number(candlesA[i].openTime),
        value: this.correlationByPrices(seriesA, seriesB),
      });
    }

    return result;
  }

  /**
   * Вычисляет логарифмические доходности по массиву свечей
   */
  private getLogReturns(candles: TIndicatorShortCandle[]) {
    const returns: number[] = [];

    for (let i = 1; i < candles.length; i++) {
      const prev = candles[i - 1].close;
      const curr = candles[i].close;

      // Защита от деления на 0 и отрицательных цен
      if (prev > 0 && curr > 0) {
        returns.push(Math.log(curr / prev));
      } else {
        returns.push(0);
      }
    }
    return returns;
  }

  /**
   * Корреляция по доходностям
   */
  public correlationByReturns(
    candlesA: TIndicatorShortCandle[],
    candlesB: TIndicatorShortCandle[],
  ) {
    const minLength = Math.min(candlesA.length, candlesB.length);
    if (minLength < 2) {
      console.warn(
        'PearsonCorrelation: prices series have less than 2 observations:',
        candlesA.length,
        candlesB.length,
      );

      return null;
    }

    const returnsA = this.getLogReturns(candlesA.slice(-minLength));
    const returnsB = this.getLogReturns(candlesB.slice(-minLength));

    // returnsA и returnsB на 1 короче, поэтому подравниваем
    const finalLength = Math.min(returnsA.length, returnsB.length);

    return this.calculate(returnsA.slice(0, finalLength), returnsB.slice(0, finalLength));
  }

  /**
   * Скользящая корреляция по доходностям
   */
  public rollingCorrelationByReturns(
    candlesA: TIndicatorShortCandle[],
    candlesB: TIndicatorShortCandle[],
    window: number = 100,
  ) {
    const minLength = Math.min(candlesA.length, candlesB.length);

    if (minLength < window + 1) {
      return [];
    }

    const returnsA = this.getLogReturns(candlesA.slice(0, minLength));
    const returnsB = this.getLogReturns(candlesB.slice(0, minLength));

    const result: { timestamp: number; value: number | null }[] = [];
    for (let i = window; i < returnsA.length && i < returnsB.length; i++) {
      const windowA = returnsA.slice(i - window, i);
      const windowB = returnsB.slice(i - window, i);

      result.push({
        timestamp: Number(candlesA[i + 1].openTime), // +1, т.к. returns на 1 меньше
        value: this.calculate(windowA, windowB),
      });
    }
    return result;
  }
}

File: src/server/trading/indicators/PearsonCorrelation/PearsonCorrelation.py
#!/usr/bin/env python3
# coding: utf-8
"""
CLI + worker-режим PearsonCorrelation
  $ python PearsonCorrelation.py   → worker (stdin/stdout)
  $ python PearsonCorrelation.py correlation_by_prices '{"candlesA":[…] …}'
"""
import json, sys, math, numpy as np
# ----------  существующие вспомогательные функции -----------
def _prices_from_candles(candles):           # … (как было)
    return [ float(c['close']) for c in candles ]

def _get_log_returns(prices):
    out = []
    for i in range(1, len(prices)):
        p, q = prices[i-1], prices[i]
        out.append(math.log(q/p) if p>0 and q>0 else 0.0)
    return out

def _corr(x, y):
    if len(x) < 2 or len(y) < 2: return 0.0
    return float(np.corrcoef(x[:len(y)], y[:len(x)])[0, 1])

# ----------  публичные API-функции -----------
def correlation_by_prices(candlesA, candlesB):
    return _corr(_prices_from_candles(candlesA), _prices_from_candles(candlesB))

def correlation_by_returns(candlesA, candlesB):
    n = min(len(candlesA), len(candlesB))
    if n < 2: return 0.0
    rA = _get_log_returns(_prices_from_candles(candlesA[:n]))
    rB = _get_log_returns(_prices_from_candles(candlesB[:n]))
    return _corr(rA, rB)

def rolling_correlation_by_prices(candlesA, candlesB, window=100):
    n = min(len(candlesA), len(candlesB))
    if n < window: return []
    out = []
    for i in range(window, n):
        out.append({
            "timestamp": int(candlesA[i]["openTime"]),
            "value": correlation_by_prices(candlesA[i-window:i], candlesB[i-window:i])
        })
    return out

def rolling_correlation_by_returns(candlesA, candlesB, window=100):
    n = min(len(candlesA), len(candlesB))
    if n < window+1: return []
    out=[]
    for i in range(window, n-1):
        out.append({
            "timestamp": int(candlesA[i+1]["openTime"]),
            "value": correlation_by_returns(candlesA[i-window:i+1], candlesB[i-window:i+1])
        })
    return out

# ----------  диспетчер -----------
_FUNCS = {
    "correlation_by_prices": correlation_by_prices,
    "correlation_by_returns": correlation_by_returns,
    "rolling_correlation_by_prices": rolling_correlation_by_prices,
    "rolling_correlation_by_returns": rolling_correlation_by_returns,
}

def _dispatch(req):
    f = _FUNCS[req["func"]]
    payload = req.get("payload", {})
    return f(**payload)

# ----------  CLI  /  WORKER -----------
def _write(obj):  sys.stdout.write(json.dumps(obj)+'\n'); sys.stdout.flush()

if __name__ == "__main__":
    if len(sys.argv) > 2:
        # однократный CLI-вызов
        func = sys.argv[1]
        payload = json.loads(sys.argv[2])
        _write(_FUNCS[func](**payload))
    else:
        # режим воркера
        for line in sys.stdin:
            try:
                req = json.loads(line)
                _write({"id": req["id"], "result": _dispatch(req)})
            except Exception as e:
                _write({"id": req.get("id"), "error": str(e)})
